moveit_core/collision_detection_bullet/include/moveit/collision_detection_bullet/tesseract/basic_types.h

```diff
@@ -0,0 +1,109 @@
+ * @par
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef TESSERACT_CORE_BASIC_TYPES_H
+#define TESSERACT_CORE_BASIC_TYPES_H
+
+#include <Eigen/Core>
+#include <Eigen/Geometry>
+#include <Eigen/StdVector>
+#include <geometric_shapes/shapes.h>
+#include <unordered_map>
+#include <vector>
+#include <memory>
+#include <functional>
+#include <map>
+#include <moveit/collision_detection/collision_common.h>
+
+namespace collision_detection_bullet
+{
+template <typename T>
+using AlignedVector = std::vector<T, Eigen::aligned_allocator<T>>;
+
+template <typename Key, typename Value>
+using AlignedMap = std::map<Key, Value, std::less<Key>, Eigen::aligned_allocator<std::pair<const Key, Value>>>;
+
+template <typename Key, typename Value>
+using AlignedUnorderedMap = std::unordered_map<Key, Value, std::hash<Key>, std::equal_to<Key>,
+                                               Eigen::aligned_allocator<std::pair<const Key, Value>>>;
+
+/** @brief Checks if contact is allowed or not. Returns true if yes, otherwise false.
+ *
+ * The order of strings should not matter, the function should handled by the function. */
+typedef std::function<bool(const std::string&, const std::string&, const collision_detection::AllowedCollisionMatrix*)>
+    IsContactAllowedFn;
+
+enum class CollisionObjectType
+{
+  UseShapeType = 0, /**< @brief Infer the type from the type specified in the shapes::Shape class */
+
+  // all of the following convert the meshes to custom collision objects
+  ConvexHull = 1,  /**< @brief Use the mesh in shapes::Shape but make it a convex hulls collision object (if not convex
+                      it will be converted) */
+  MultiSphere = 2, /**< @brief Use the mesh and represent it by multiple spheres collision object */
+  SDF = 3          /**< @brief Use the mesh and rpresent it by a signed distance fields collision object */
+};
+
+/// Contact test data and query results information
+struct ContactTestData
+{
+  EIGEN_MAKE_ALIGNED_OPERATOR_NEW
+
+  ContactTestData(const std::vector<std::string>& active, const double& contact_distance, const IsContactAllowedFn& fn,
+                  collision_detection::CollisionResult& res, const collision_detection::CollisionRequest& req,
+                  const collision_detection::AllowedCollisionMatrix* acm)
+    : active(active)
+    , contact_distance(contact_distance)
+    , fn(fn)
+    , acm(acm)
+    , res(res)
+    , req(req)
+    , done(false)
+    , pair_done(false)
+  {
+  }
+
+  const std::vector<std::string>& active;
+
+  /** \brief If after a positive broadphase check the distance is below this threshold, a contact is added. */
+  const double& contact_distance;
```

Comment:
Shouldn't this be called `broadphase_threshold` then (according to the comment)?
A contact should only be added if there is a real contact...

Reply:
No, this is not the broadphase threshold, this is another parameter specific to each collision object. For a broadphase candidate pair, their distance is calculated and if its below the `contact_distance` the distance and all other information is added as a result. In Bullet this is regarded as a contact. Note that contact does not necessarily refer to a real contact (distance < 0 is possible).

Contrary to FCL, Bullet does not make a difference between distance and collision requests which in my opinion is the cleaner solution with less duplicate code (the FCL MoveIt implementation has plenty of duplicate code because of this distinction). In Bullet, distance requests are simply treated as contacts with the only difference that the contact depth/distance is positive. So a contact is not necessarily a real collision. This wording is used throughout the Bullet Library.

Reply:
If I get you correct, Bullet performs the detailed collision/distance check only if the broadphase check passed. So, if two objects fail the broadphase check (e.g. because they are nicely aligned and their axes-aligned bounding boxes don't overlap), their distance will not be reported?

Reply:
No, they won't. But you can easily expand the AABB of each object by a distance (this is a parameter) and then they will be reported. So I could expand each AABB by 5 cm and then I will get for sure all objects which are closer than 5 cm.